crates/ecs/src/lib.rs

```diff
@@ -508,14 +533,48 @@ pub struct World {
 type ReadComponentVec<'a, ComponentType> = Option<RwLockReadGuard<'a, Vec<Option<ComponentType>>>>;
 type WriteComponentVec<'a, ComponentType> =
     Option<RwLockWriteGuard<'a, Vec<Option<ComponentType>>>>;
-type TargetArchSourceTargetIDs = (Option<ArchetypeIndex>, (Vec<TypeId>, Vec<TypeId>));
+type TargetArchSourceTargetIDs = (Option<ArchetypeIndex>, Vec<TypeId>, Vec<TypeId>);
+
+enum ArchetypeMutation {
+    Removal(Entity),
+    Addition(Entity),
+}
 
 impl World {
+    /// Returns the archetype index of the archetype that contains all
+    /// components types that the entity is tied to -+ the generic type
+    /// supplied to the function given "removal flag" status.
+    /// None is returned if no archetype containing only the sought after component types exist.
+    /// The type Ids contained within the archetype the entity is existing in and the type ids for
+    /// sought archetype are also returned.
+    ///
+    /// Set removal to true or false depending on if the supplied generic type should be added to or
+    /// removed from the specifie entity.
+    /// For example call world.find_target_archetype::<u32>(entity, true)
+    /// to fetch the archetype index of the archetype
+    /// containing all component types except u32 that the entity is tied to.
+    ///
+    /// world.find_target_archetype::<u32>(entity, false) would return the archetype index of the
+    /// archetype containing all component types that the entity is tied to with the addition
+    /// of u32.
     fn find_target_archetype<ComponentType: Debug + Send + Sync + 'static>(
-        &mut self,
-        entity: Entity,
-        removal: bool,
+        &self,
+        mutation: ArchetypeMutation,
     ) -> WorldResult<TargetArchSourceTargetIDs> {
+        let entity;
+        let removal;
+
+        match mutation {
+            ArchetypeMutation::Addition(nested_entity) => {
+                entity = nested_entity;
+                removal = false;
+            }
+            ArchetypeMutation::Removal(nested_entity) => {
+                entity = nested_entity;
+                removal = true;
+            }
+        }
```

Comment:
While this works, it's not how you usually handle enums that contain nested data.

You've got a match-statement later in this function which, together with the code I've marked here, can be replaced with the following:

```rust
match mutation {
    ArchetypeMutation::Addition(entity) => {
        // todo: in here you can place the code you previously had under the match removal false-branch.
        // todo: in here you've got access to `entity` and you don't need the bool `removal`
    }
    ArchetypeMutation::Removal(entity) => {
        // todo: in here you can place the code you previously had under the match removal true-branch.
        // todo: in here you've got access to `entity` and you don't need the bool `removal`
    }
}
```